src/serac/numerics/functional/tensor.hpp

```diff
@@ -1170,58 +1172,147 @@ SERAC_HOST_DEVICE bool is_symmetric_and_positive_definite(tensor<double, 3, 3> A
 }
 
 /**
- * @brief Solves Ax = b for x using Gaussian elimination with partial pivoting
- * @param[in] A The coefficient matrix A
- * @param[in] b The righthand side vector b
- * @note @a A and @a b are by-value as they are mutated as part of the elimination
+ * @brief Compute LU factorization of a matrix with partial pivoting
+ * @param[in] A The matrix to factorize
+ * @return x tuple of [P, L, U], with
+ * L the lower triangular factor
+ * U the upper triangular factor
+ * P an array indicating the row permutations due to pivoting.
+ * Row i of A was exchanged with row P[i].
  */
 template <typename T, int n>
-SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(tensor<T, n, n> A, const tensor<T, n> b)
+SERAC_HOST_DEVICE constexpr auto lu(const tensor<T, n, n>& A)
 {
+  // BT question for Sam Mish: Why not use the standard library abs?
   constexpr auto abs  = [](double x) { return (x < 0) ? -x : x; };
   constexpr auto swap = [](auto& x, auto& y) {
     auto tmp = x;
     x        = y;
     y        = tmp;
   };
 
-  tensor<double, n> x{};
+  auto           U = A;
+  auto           L = DenseIdentity<n>();
+  tensor<int, n> P(make_tensor<n>([](auto i) { return i; }));
 
   for (int i = 0; i < n; i++) {
     // Search for maximum in this column
-    double max_val = abs(A[i][i]);
+    double max_val = abs(U[i][i]);
 
     int max_row = i;
     for (int j = i + 1; j < n; j++) {
-      if (abs(A[j][i]) > max_val) {
-        max_val = abs(A[j][i]);
+      if (abs(U[j][i]) > max_val) {
+        max_val = abs(U[j][i]);
         max_row = j;
       }
     }
 
-    swap(b[max_row], b[i]);
-    swap(A[max_row], A[i]);
+    swap(P[max_row], P[i]);
+    swap(U[max_row], U[i]);
+  }
 
-    // zero entries below in this column
+  for (int i = 0; i < n; i++) {
+    // zero entries below in this column in U
+    // and fill in L entries
     for (int j = i + 1; j < n; j++) {
-      double c = -A[j][i] / A[i][i];
-      A[j] += c * A[i];
-      b[j] += c * b[i];
-      A[j][i] = 0;
+      double c = U[j][i] / U[i][i];
```

Comment:
using `double` explicitly might make this do the wrong thing when `T` is a dual type

Reply:
The dual type case shouldn't happen, if I've done things right, because there is a specific implementation to handle that case. But your point stands, this will do the wrong thing if T is another type like quad precision, say. Will `T c = U[j][i] / U[i][i];` suffice?

Reply:
Or I guess `auto` would be best. Your comment applies to line 1200 as well.

Reply:
I wrote a test in https://github.com/LLNL/serac/pull/691/commits/b0570fe2cd38e112980fdcff2d14622bd8740021 that verifies that the LU factorization code fails to compile when instantiated with dual numbers, and then, with your help, we modified the implementation to handle this case. The test now passes. It checks the value and the gradient produced by LU.